src/Qir/Tests/QIR-dynamic/qsharp/qir-test-assert.qs

```diff
@@ -1,15 +1,235 @@
+    open Microsoft.Quantum.Intrinsic;
+
+    // |0> -> |1>
+    @EntryPoint()
+    operation AsrtMeasAlloc1OKTest() : Unit {
+        use qubit = Qubit();    // |0>
+
+        AssertMeasurement(           [PauliZ], [qubit], Zero,      "0: Newly allocated qubit must be in the |0> state.");
+        AssertMeasurementProbability([PauliZ], [qubit], Zero, 1.0, "1: Newly allocated qubit must be in the |0> state.", 1e-10);
+
+        X(qubit);   // |0> -> |1>
+
+        AssertMeasurement(           [PauliZ], [qubit], One,      "2: Newly allocated qubit after X() must be in the |1> state.");
+        AssertMeasurementProbability([PauliZ], [qubit], One, 1.0, "3: Newly allocated qubit after X() must be in the |1> state.", 1e-10);
+    }
+
+    // Fail: Expecting |1>, should have expected |0>
+    @EntryPoint()
+    operation AsrtMeasAlloc1ExcTest() : Unit {
+        use qubit = Qubit();    // |0>
+
+        // Must fail:
+        AssertMeasurement([PauliZ], [qubit], One, "Newly allocated qubit must be in the |0> state.");
+    }
+
+    // Fail: Expecting |1>, should have expected |0>
+    @EntryPoint()
+    operation AsrtMeasProbAlloc1ExcTest() : Unit {
+        use qubit = Qubit();    // |0>
+
+        // Must fail:
+        AssertMeasurementProbability([PauliZ], [qubit], One, 1.0, "Newly allocated qubit must be in the |0> state.", 1e-10);
+    }
+
+    @EntryPoint()
+    operation AsrtMeasProbAlloc1HalfProbTest() : Unit {     //!
+        use qubit = Qubit();    // |0>
+
+        // Measuring in PauliX: 50% |0>, 50% |1>
+        AssertMeasurementProbability([PauliX], [qubit], Zero, 0.5, "0: Newly allocated qubit must be in the |0> state.", 1e-3);
+        AssertMeasurementProbability([PauliX], [qubit],  One, 0.5, "1: Newly allocated qubit must be in the |0> state.", 1e-3);
+        // Measuring in PauliY: 50% |0>, 50% |1>
+        AssertMeasurementProbability([PauliY], [qubit], Zero, 0.5, "2: Newly allocated qubit must be in the |0> state.", 1e-3);
+        AssertMeasurementProbability([PauliY], [qubit],  One, 0.5, "3: Newly allocated qubit must be in the |0> state.", 1e-3);
+    }
+
+    // |+>, |->
+    @EntryPoint()
+    operation AsrtMeasProbAllocPlusMinusTest() : Unit {
+        let str = "Newly allocated qubit after H() must be in the |+> state";
+
+        use qubit = Qubit();    // |0>
+        H(qubit);               // |+>
+        AssertMeasurement(           [PauliX], [qubit], Zero, str);
+        // 50% probability in other Pauli bases:
+        AssertMeasurementProbability([PauliZ], [qubit], Zero, 0.5, str, 1e-10);
+        AssertMeasurementProbability([PauliZ], [qubit],  One, 0.5, str, 1e-10);
+        AssertMeasurementProbability([PauliY], [qubit], Zero, 0.5, str, 1e-10);
+        AssertMeasurementProbability([PauliY], [qubit],  One, 0.5, str, 1e-10);
+
+        let str2 = "Newly allocated qubit after x() followed by H() must be in the |-> state";
+        H(qubit);   // Back to |0>
+        X(qubit);   // |1>
+        H(qubit);   // |->
```

Comment:
𝐻𝑋𝐻 = 𝑍, such that the calls to `H` and `X` can be greatly simplified here.
```suggestion
        Z(qubit);
```

To see why 𝑍|+⟩ = |−⟩, it helps to write it out in the computational basis using Dirac notation: 𝑍|+⟩ = 𝑍(|0⟩ + |1⟩) / √2 = (𝑍|0⟩ + 𝑍|1⟩) / √2 = (|0⟩ − |1⟩) / √2 = |−⟩, where we used that 𝑍|0⟩ = |0⟩ and 𝑍|1⟩ = −|1⟩. Put differently, 𝑍 does the same thing to the 𝑋-basis (|+⟩ and |−⟩) that 𝑋 does to the 𝑍-basis (|0⟩ and |1⟩), 𝑍|+⟩ = |−⟩ and 𝑍|−⟩ = |+⟩.

Reply:
Thanks for the useful info. I have updated my records.
But after adding `within`+`apply` the sequence has stopped being applicable (the fragment `H(qubit);   // Back to |0>` is done implicitly in the end of `apply`).